addons/website_sale/controllers/main.py

```diff
@@ -1225,3 +1225,56 @@ def products_autocomplete(self, term, options={}, **kwargs):
+    # Products Recently Viewed
+    # --------------------------------------------------------------------------
+    @http.route('/shop/products/recently_viewed', type='json', auth='public', website=True)
+    def products_recently_viewed(self, **kwargs):
+        """
+        Returns list of recently viewed products according to current user and product options
+        """
+        excluded_ids = []
+        current_sale_order = request.session.get('sale_order_id')
+        if current_sale_order:
+            order_lines = request.env['sale.order.line'].sudo().search([('order_id', '=', current_sale_order)])
+            excluded_ids += order_lines.mapped('product_id.id')
+        if not request.env.user._is_public():
+            partner_id = request.env.user.partner_id
+            visitor_sudo = request.env['website.visitor'].sudo().with_context(active_test=False).search([('user_partner_id', '=', partner_id.id)])
+        else:
+            visitor_sudo = request.env['website.visitor']._get_visitor_from_request()
+        if visitor_sudo:
+            length = 12 + len(excluded_ids)
+            products = visitor_sudo.website_track_ids.product_id[:length]
+            if excluded_ids:
+                products = products.filtered(lambda r: r.id not in excluded_ids)[:12]
+            if products:
+                fields = ['id', 'name', 'website_url']
+                res = {
+                    'products': products.read(fields),
+                }
+
+                FieldMonetary = request.env['ir.qweb.field.monetary']
+                monetary_options = {
+                    'display_currency': request.website.get_current_pricelist().currency_id,
+                }
+                rating = request.website.viewref('website_sale.product_comment').active
+                for res_product, product in zip(res['products'], products):
+                    combination_info = product._get_combination_info_variant()
+                    res_product.update(combination_info)
+                    res_product['list_price'] = FieldMonetary.value_to_html(res_product['list_price'], monetary_options)
+                    res_product['price'] = FieldMonetary.value_to_html(res_product['price'], monetary_options)
+                    if rating:
+                        res_product['rating'] = request.env["ir.ui.view"].render_template('website_rating.rating_widget_stars_static', values={
+                            'rating_avg': product.rating_avg,
+                            'rating_count': product.rating_count,
+                        })
+
+                return res
+        return {}
+
+    @http.route('/shop/products/recently_viewed_update', type='json', auth='public', website=True)
+    def products_recently_viewed_update(self, product_id, **kwargs):
+        visitor = request.env['website.visitor']._get_visitor_from_request()
+        return request.env['website.visitor']._add_viewed_product(None, product_id, visitor)
```

Comment:
Will crash if visitor is null as add_tracking will try to create one visitor and set cookie on a null object !! 
If the visitor did not go on a tracked page : no visitor created. 
I think you made half of work here (because in js controller, you set the cookie if a new visitor_sign is returned.)
Also, I think _add_viewed_product should not be a model method but should work directly on the record.
response param in _add_viewed_product is useless

At the end, this should look something like : 
```
ret = {}
visitor = request.env['website.visitor']._get_visitor_from_request()
if not visitor:
    visitor =  create_visitor()
    ret['visitor_sign'] = get_visitor_sign()
visitor._add_viewed_product(product_id)
return ret
```

Reply:
Thx for this, forgot it when I rebased on your last changes...